lighthouse-core/scripts/i18n/collect-strings.js

```diff
@@ -30,18 +28,337 @@ const ignoredPathComponents = [
+  const parts = icu.message.split(
+    /\{(\w+), (\w+), (\w+)\}/g);
+  icu.message = '';
+  let idx = 0;
+
+  while (parts.length) {
+    // Seperate out the match into parts.
+    const [preambleText, rawName, format, formatType] = parts.splice(0, 4);
+    icu.message += preambleText;
+
+    if (!rawName || !format || !formatType) continue;
+    // Check that custom-formatted ICU not using non-supported format ex:
+    // * using a second arg anything other than "number"
+    // * using a third arg that is not millis, secs, bytes, %, or extended %
+    if (!format.match(/^number$/)) {
+      throw Error(`Unsupported custom-formatted ICU format var "${format}" in message "${icu.message}"`);
+    }
+    if (!formatType.match(/milliseconds|seconds|bytes|percent|extendedPercent/)) {
+      throw Error(`Unsupported custom-formatted ICU type var "${formatType}" in message "${icu.message}"`);
+    }
+
+    // Append ICU replacements if there are any.
+    const placeholderName = `CUSTOM_ICU_${idx++}`;
+    icu.message += `$${placeholderName}$`;
+    let example;
+
+    // Make some good examples.
+    switch (formatType) {
+      case 'seconds':
+        example = '2.4';
+        break;
+      case 'percent':
+        example = '54.6%';
+        break;
+      case 'extendedPercent':
+        example = '37.92%';
+        break;
+      case 'milliseconds':
+      case 'bytes':
+        example = '499';
+        break;
+      default:
+        // This shouldn't be possible, but if the above formatType regex fails, this is fallback.
+        throw Error('Unknown formatType');
+    }
+
+    icu.placeholders[placeholderName] = {
+      content: `{${rawName}, number, ${formatType}}`,
+      example,
+    };
+  }
+}
+
+/**
+ * Add examples for direct ICU replacement.
+ *
+ * @param {ICUMessageDefn} icu
+ * @param {Record<string, string>} examples
+ */
+function _processPlaceholderDirectIcu(icu, examples) {
+  let tempMessage = icu.message;
+  let idx = 0;
+  const findIcu = /\{(\w+)\}/g;
+
+  let matches;
+  // Make sure all ICU vars have examples
+  while ((matches = findIcu.exec(tempMessage)) !== null) {
+    const varName = matches[1];
+    if (!examples[varName]) {
+      throw Error(`Variable '${varName}' is missing example comment in message "${tempMessage}"`);
+    }
+  }
+
+  for (const [key, value] of Object.entries(examples)) {
+    // Make sure all examples have ICU vars
+    if (!icu.message.includes(`{${key}}`)) {
+      throw Error(`Example '${key}' provided, but has not corresponding ICU replacement in message "${icu.message}"`);
+    }
+    const eName = `ICU_${idx++}`;
+    tempMessage = tempMessage.replace(`{${key}}`, `$${eName}$`);
+
+    icu.placeholders[eName] = {
+      content: `{${key}}`,
+      example: value,
+    };
+  }
+  icu.message = tempMessage;
+}
+
+/**
+ * Take a series of messages and apply ĥât̂ markers to the translatable portions
+ * of the text.  Used to generate `en-XL` locale to debug i18n strings.
```

Comment:
maybe add a note to the comment that we do this in ctc format, not lhl, so that literal placeholders are not transformed? (I was just about to suggest moving this and had to think it through :)